drivers/can/socket_can_generic.h

```diff
@@ -43,36 +44,107 @@ static inline void socket_can_iface_init(struct net_if *iface)
 	LOG_DBG("Init CAN interface %p dev %p", iface, dev);
 }
 
-static inline void tx_irq_callback(u32_t error_flags, void *arg)
+static inline int socket_can_tx_state_to_frame(enum can_state state)
 {
-	char *caller_str = (char *)arg;
-	if (error_flags) {
-		LOG_DBG("TX error from %s! error-code: %d",
-			caller_str, error_flags);
+	int err = 0;
+
+	switch (state) {
+	case CAN_ERROR_ACTIVE:
+		err = CAN_ERR_CRTL_ACTIVE;
+		break;
+	case CAN_ERROR_PASSIVE:
+		err = CAN_ERR_CRTL_TX_PASSIVE;
+		break;
+	default:
+		err = 0;
 	}
+	return err;
 }
 
+static inline int socket_can_rx_state_to_frame(enum can_state state)
+{
+	int err = 0;
+
+	switch (state) {
+	case CAN_ERROR_ACTIVE:
+		err = CAN_ERR_CRTL_ACTIVE;
+		break;
+	case CAN_ERROR_PASSIVE:
+		err = CAN_ERR_CRTL_RX_PASSIVE;
+		break;
+	default:
+		err = 0;
+	}
+	return err;
+}
+
+static inline void socket_can_change_state(struct socket_can_context *ctx,
+						enum can_state tx_state,
+						enum can_state rx_state)
+{
+	enum can_state new_state = MAX(tx_state, rx_state);
+
+	if (new_state != ctx->state) {
+		struct can_frame msg = {
+			.can_id = CAN_ERR_FLAG,
+			.can_dlc = CAN_ERR_DLC
+		};
+
+		LOG_DBG("New error state: %d", new_state);
+
+		if (new_state == CAN_BUS_OFF) {
+			msg.can_id |= CAN_ERR_BUSOFF;
+		} else {
+			msg.can_id |= CAN_ERR_CRTL;
+			msg.data[1] |= tx_state >= rx_state ?
+					socket_can_tx_state_to_frame(tx_state) :
+					0;
+			msg.data[1] |= tx_state <= rx_state ?
+					socket_can_rx_state_to_frame(rx_state) :
+					0;
+		}
+
+		ctx->state = new_state;
+
+		k_msgq_put(ctx->msgq, &msg, K_NO_WAIT);
+	}
+}
+
+
 /* This is called by net_if.c when packet is about to be sent */
 static inline int socket_can_send(struct device *dev, struct net_pkt *pkt)
 {
 	struct socket_can_context *socket_context = dev->driver_data;
+	struct zcan_frame zframe;
 	int ret;
 
 	if (net_pkt_family(pkt) != AF_CAN) {
 		return -EPFNOSUPPORT;
 	}
 
-	ret = can_send(socket_context->can_dev,
-		       (struct zcan_frame *)pkt->frags->data,
-		       SEND_TIMEOUT, tx_irq_callback, "socket_can_send");
+	can_copy_frame_to_zframe((const struct can_frame *)pkt->frags->data,
+								&zframe);
+
+	ret = can_send(socket_context->can_dev, (struct zcan_frame *)&zframe,
```

Comment:
You are changing the send call to be blocking. Is this intentional?

Reply:
Unintentional yes,I will fix this. 
I find it a bit confusing, providing a timeout and no callback will trigger a blocking call.